File: apps/detection/src/secret_scanner.py
"""
Iron Gate Secret Scanner

Detects credentials, API keys, tokens, and other secrets in text.
Ported from the TypeScript version (apps/api/src/proxy/secretScanner.ts).
"""

import re
from dataclasses import dataclass


@dataclass
class DetectedSecret:
    type: str
    text: str
    start: int
    end: int
    confidence: float


class SecretScanner:
    """
    Regex-based secret scanner that detects API keys, cloud credentials,
    database URIs, auth tokens, private keys, and other sensitive secrets.
    """

    PATTERNS = {
        'API_KEY': [
            (r'sk-[a-zA-Z0-9]{20,}', 0.95),           # OpenAI
            (r'sk_live_[a-zA-Z0-9]{24,}', 0.95),       # Stripe
            (r'sk_test_[a-zA-Z0-9]{24,}', 0.90),       # Stripe test
            (r'ghp_[a-zA-Z0-9]{36}', 0.95),             # GitHub PAT
            (r'xoxb-[0-9]+-[a-zA-Z0-9]+', 0.90),       # Slack bot
            (r'SG\.[a-zA-Z0-9_-]{22}\.[a-zA-Z0-9_-]{43}', 0.95),  # SendGrid
            (r'sk-ant-[a-zA-Z0-9_-]{40,}', 0.95),      # Anthropic
        ],
        'AWS_CREDENTIAL': [
            (r'AKIA[0-9A-Z]{16}', 0.95),
            (r'ASIA[0-9A-Z]{16}', 0.90),
        ],
        'GCP_CREDENTIAL': [
            (r'AIza[0-9A-Za-z_-]{35}', 0.90),
        ],
        'DATABASE_URI': [
            (r'(?:postgres(?:ql)?|mysql|mongodb|redis)://\S+', 0.95),
        ],
        'AUTH_TOKEN': [
            (r'eyJ[a-zA-Z0-9_-]+\.eyJ[a-zA-Z0-9_-]+\.[a-zA-Z0-9_-]+', 0.90),  # JWT
        ],
        'PRIVATE_KEY': [
            (r'-----BEGIN (?:RSA |EC |DSA )?PRIVATE KEY-----', 0.99),
        ],
        'AZURE_CREDENTIAL': [
            (r'DefaultEndpointsProtocol=https;AccountName=\S+', 0.90),
        ],
    }

    def detect(self, text: str) -> list[DetectedSecret]:
        """
        Scan text for known secret patterns.

        Returns a list of DetectedSecret instances with type, matched text,
        character offsets, and confidence score.
        """
        results = []
        for entity_type, patterns in self.PATTERNS.items():
            for pattern, confidence in patterns:
                for match in re.finditer(pattern, text):
                    results.append(DetectedSecret(
                        type=entity_type,
                        text=match.group(),
                        start=match.start(),
                        end=match.end(),
                        confidence=confidence,
                    ))
        return results

File: apps/extension/src/worker/detectors/secret-scanner.ts
// ============================================================================
// Iron Gate — Extension Secret Scanner
// ============================================================================
// Client-side secret detection mirroring the API's secret-scanner.ts.
// Runs in the service worker for real-time prompt scanning.
// ============================================================================

interface DetectedSecret {
  type: string;
  text: string;
  start: number;
  end: number;
  confidence: number;
  source: 'regex';
}

interface SecretPattern {
  type: string;
  patterns: RegExp[];
  confidence: number;
}

const SECRET_PATTERNS: SecretPattern[] = [
  {
    type: 'API_KEY',
    patterns: [
      /\bsk-[a-zA-Z0-9]{20,}\b/g,                          // OpenAI-style
      /\bsk_live_[a-zA-Z0-9]{24,}\b/g,                      // Stripe-style
      /\bsk-ant-[a-zA-Z0-9\-]{20,}\b/g,                     // Anthropic-style
      /\bghp_[a-zA-Z0-9]{36}\b/g,                           // GitHub PAT
      /\bgho_[a-zA-Z0-9]{36}\b/g,                           // GitHub OAuth
      /\bghs_[a-zA-Z0-9]{36}\b/g,                           // GitHub App
      /\bxoxb-[0-9]+-[a-zA-Z0-9]+\b/g,                      // Slack bot token
      /\bxoxp-[0-9]+-[a-zA-Z0-9]+\b/g,                      // Slack user token
      /\bSG\.[a-zA-Z0-9_-]{22}\.[a-zA-Z0-9_-]{43}\b/g,     // SendGrid
    ],
    confidence: 0.95,
  },
  {
    type: 'AWS_CREDENTIAL',
    patterns: [
      /\bAKIA[0-9A-Z]{16}\b/g,                              // AWS Access Key ID
      /\bASIA[0-9A-Z]{16}\b/g,                              // AWS Temp Access Key
    ],
    confidence: 0.95,
  },
  {
    type: 'GCP_CREDENTIAL',
    patterns: [
      /\bAIza[0-9A-Za-z_-]{35}\b/g,                         // GCP API Key
    ],
    confidence: 0.9,
  },
  {
    type: 'DATABASE_URI',
    patterns: [
      /\b(?:postgres|postgresql|mysql|mongodb|mongodb\+srv|redis):\/\/[^\s"']+/g,
    ],
    confidence: 0.95,
  },
  {
    type: 'AUTH_TOKEN',
    patterns: [
      /\beyJ[a-zA-Z0-9_-]{10,}\.eyJ[a-zA-Z0-9_-]{10,}\.[a-zA-Z0-9_-]{10,}\b/g,  // JWT
    ],
    confidence: 0.9,
  },
  {
    type: 'PRIVATE_KEY',
    patterns: [
      /-----BEGIN (?:RSA |EC |DSA |OPENSSH )?PRIVATE KEY-----/g,
    ],
    confidence: 0.99,
  },
];

/**
 * Scan text for secrets (API keys, tokens, credentials).
 * Mirrors the API server's secret-scanner.ts for client-side detection.
 */
export function scanForSecrets(text: string): DetectedSecret[] {
  const secrets: DetectedSecret[] = [];
  const seen = new Set<string>();

  for (const { type, patterns, confidence } of SECRET_PATTERNS) {
    for (const pattern of patterns) {
      pattern.lastIndex = 0;
      let match: RegExpExecArray | null;

      while ((match = pattern.exec(text)) !== null) {
        const key = `${match.index}-${match.index + match[0].length}-${type}`;
        if (!seen.has(key)) {
          seen.add(key);
          secrets.push({
            type,
            text: match[0],
            start: match.index,
            end: match.index + match[0].length,
            confidence,
            source: 'regex',
          });
        }
      }
    }
  }

  return secrets;
}
